
<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<!--
  #%L
  BDBM GUI
  %%
  Copyright (C) 2014 - 2017 Miguel Reboiro-Jato, Critina P. Vieira, Hugo López-Fdez, Noé Vázquez González, Florentino Fdez-Riverola and Jorge Vieira
  %%
  This program is free software: you can redistribute it and/or modify
  it under the terms of the GNU General Public License as
  published by the Free Software Foundation, either version 3 of the
  License, or (at your option) any later version.
  
  This program is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU General Public License for more details.
  
  You should have received a copy of the GNU General Public
  License along with this program.  If not, see
  <http://www.gnu.org/licenses/gpl-3.0.html>.
  #L%
  -->

<html>
<head>
  <title>Refine Annotation</title>
</head>
<body>
  <h1><i>Refine Annotation</i></h1>
  <p>Depending on the number and location of the differences found between 
    the reference CDS and the target sequences, the <i>Splign-Compart (NCBI)</i>
    and the <i>ProSplign-Compart (NCBI)</i> options do not always provide a 
    complete CDS annotation. Nevertheless, if all exon-intron splice junctions
    are covered in the partial CDS annotation it may be possible to obtain a
    complete CDS annotation by combining the results produced by the
    <i>Splign-Compart (NCBI)</i> and the <i>ProSplign-Compart (NCBI)</i> 
    options and information on putative open reading frames generated by 
    the <i>getorf</i> application. When using the <i>Refine Annotation</i>
    option the user must provide the name of the <i>FASTA</i> file 
    generated by the <i>Splign-Compart (NCBI)</i> or the 
    <i>ProSplign-Compart (NCBI)</i> options, as well as a <i>FASTA</i> 
    file with approximate genome region where the gene is located (for 
    instance the file with the Fasta extension that is saved by the
    <i>ProSplign-Compart (NCBI)</i> option in the <i>/Export Files/nucleotides</i> folder, 
    located in the specified repository folder. Moreover, it must specify 
    the size of the region used to determine if there is an overlap, as
    well as the minimum and maximum size of the open reading frames to be
    reported by the <i>getorf</i> application.
  </p>
  <p>The <i>Refine Annotation</i> option will automatically perform the following
    steps:
  </p>
  <ol>
    <li>Get all open reading frames (between STOP codons) for the provided genome 
      region (only the plus strand is considered thus it is important to give the
      genome sequence in the proper orientation) and sort them by
      sequence size (from the longest to the shortest).
    </li>
    <li>For each sequence in the partial CDS annotation file: extract the last n 
      positions of the sequence and try to find a match in the open reading frames obtained 
      in (1); only the first match is considered; if a match is found add the piece
      of sequence found after the hit to the sequence from where the motif
      used for the search originates from.
    </li>
    <li>For all sequences obtained in (2): compare each possible pair of sequences 
      in order to extract the first n positions of the first sequence and try to find
      a match in the second sequence; if a match is found erase the sequence region 
      from the hit to the end of the target sequence and concatenate the two sequences.
    </li>
    <li>Repeat step (3) until there is no possibility of merging two sequences.</li>
    <li>Get all open reading frames (between a START and a STOP codon) for
      the provided genome region and sort them by size (from the longest to the shortest).
    </li>
    <li>For each sequence obtained in (4): extract the first n positions of the
      sequence and try to find a match in the open reading frames obtained 
      in (5); only the first match is
      considered;  if a hit is found, erase the sequence of the open reading frame showing 
      the hit from the place where the motif is found until the end and prefix the first
      sequence being processed with the sequence just obtained.
    </li>
    <li>For all sequences obtained in (6): compare each possible pair of sequences 
      in order to extract the first n positions of the first sequence and try to find
      a match in the second sequence; if a match is found erase the sequence region 
      from the hit to the end of the target sequence and concatenate the two sequences.
    </li>
    <li>Repeat step (7) until there is no possibility of merging two sequences.</li>
  </ol>
  <div align="center">
    <img src="../Images/DialogRefineAnnotation.png" border="0"/><br/>
    <strong><i>Refine annotation</i></strong>
  </div>
</body>
</html>
