<!--
  #%L
  BDBM GUI
  %%
  Copyright (C) 2014 - 2015 Miguel Reboiro-Jato, Critina P. Vieira, Hugo López-Fdez, Florentino Fdez-Riverola and Jorge Vieira
  %%
  This program is free software: you can redistribute it and/or modify
  it under the terms of the GNU General Public License as
  published by the Free Software Foundation, either version 3 of the
  License, or (at your option) any later version.
  
  This program is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU General Public License for more details.
  
  You should have received a copy of the GNU General Public
  License along with this program.  If not, see
  <http://www.gnu.org/licenses/gpl-3.0.html>.
  #L%
  -->
<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<html>
<head>
  <title>Splign-Compart (NCBI)</title>
</head>
<body>
  <h1><i>Splign-Compart (NCBI)</i></h1>
   <p>When genomes are non-annotated and genes have (or are suspected to have) introns, the <i>Splign-Compart (NCBI)</i> option under <i>Operations</i> can be used to annotate exons or genes, as long as a CDS reference sequence is available from a closely related species. How closely related the species must be depends on how fast the gene(s) in question evolve. For instance, a few highly conserved <i>Drosophila virilis</i> genes can be annotated this way using as reference <i>Drosophila melanogaster</i> CDSs (the common ancestor of the two species lived more than 40 million years ago). When using the <i>Splign-Compart (NCBI)</i> option the user must specify the FASTA file containing the genome to be annotated (<i>Genome Fasta</i>), as well as the file containing the reference CDSs (<i>CDS Fasta</i>) and the name of the output file (<i>Output name</i>). The resulting file will be saved in the /fasta/nucleotides folder that is located in the specified repository folder, and thus can be used in further operations such as Blast analyses. If the Concatenate Exons option is used then adjacent exons will be concatenated. Therefore, if an annotation is obtained for every exon of a given gene, the resulting sequence will be the complete CDS. The resulting CDS is based on the nucleotide homology to a given sequence, and thus may produce CDS sequences with lengths that are not multiple of three, if for instance, sequencing errors causing frameshifts are present in the genome to be annotated. Nevertheless, the existence of intron splicing signals at the exons 5’ and 3’ ends is taken into account. When using this option, it is advisable to use short FASTA headers, avoid the use of special characters, as well as new lines within sequences, otherwise the pipeline will likely crash. Remember that the reformatting of any FASTA file can be easily performed using the <i>BDBM Reformat Fasta</i> option under the <i>Operations</i> tab (see below). Since this is a homology based approach, it is possible to keep the stop codon in the reference sequence. When this is done, the stop codon will be likely included in the resulting annotation as well, increasing the chances of having a complete annotation, even when the previous codon differs in the reference and target sequences. The inclusion of a stop codon in the CDS annotation is also a guarantee that the CDS is indeed complete.</p>
  <div align="center">
    <p>
      <img src="../Images/DialogSplignCompart.png" border="0"/><br/>
      <strong><i>Splign-Compart (NCBI)</i></strong>
    </p>
  </div>
</body>
</html>
